src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -306,104 +357,526 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+     chunks using the larger chunk, but all corner cases need to be considered like the last chunk which might be smaller than obj_max_req_size
+     and also ranged requests where a smaller chunk is overwritten by a larger chunk size != obj_max_req_size */
+
+  uint64_t obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+  uint64_t start_part_num = 0;
+  uint64_t part_num = ofs/obj_max_req_size; //part num of ofs wrt start of the object
+  uint64_t adjusted_start_ofs = part_num*obj_max_req_size; //in case of ranged request, adjust the start offset to the beginning of a chunk/ part
+  uint64_t diff_ofs = ofs - adjusted_start_ofs; //difference between actual offset and adjusted offset
+  off_t len = (end - adjusted_start_ofs) + 1;
+  uint64_t num_parts = (len%obj_max_req_size) == 0 ? len/obj_max_req_size : (len/obj_max_req_size) + 1; //calculate num parts based on adjusted offset
+  //len_to_read is the actual length read from a part/ chunk in cache, while part_len is the length of the chunk/ part in cache 
+  uint64_t cost = 0, len_to_read = 0, part_len = 0;
+
+  aio = rgw::make_throttle(window_size, y);
+
+  ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << "obj_max_req_size " << obj_max_req_size << 
+  " num_parts " << num_parts << " adjusted_start_offset: " << adjusted_start_ofs << " len: " << len << dendl;
+
+  this->offset = ofs;
+
+  do {
+    uint64_t id = adjusted_start_ofs, read_ofs = 0; //read_ofs is the actual offset to start reading from the current part/ chunk
+    if (start_part_num == (num_parts - 1)) {
+      len_to_read = len;
+      part_len = len;
+      cost = len;
+    } else {
+      len_to_read = obj_max_req_size;
+      cost = obj_max_req_size;
+      part_len = obj_max_req_size;
+    }
+    if (start_part_num == 0) {
+      len_to_read -= diff_ofs;
+      id += diff_ofs;
+      read_ofs = diff_ofs;
+    }
+
+    ceph::bufferlist bl;
+    std::string oid_in_cache = prefix + "_" + std::to_string(adjusted_start_ofs) + "_" + std::to_string(part_len);
+    if (version.empty()) {
+      version = source->get_instance();
+    }
+
+    ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << "version stored in update method is: " << version << dendl;
+
+    ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << 
+    " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+
+    if (source->driver->get_policy_driver()->get_cache_policy()->exist_key(oid_in_cache) > 0) { 
+      // Read From Cache
+      auto completed = source->driver->get_cache_driver()->get_async(dpp, y, aio.get(), oid_in_cache, read_ofs, len_to_read, cost, id); 
+
+      source->driver->get_policy_driver()->get_cache_policy()->update(dpp, oid_in_cache, adjusted_start_ofs, part_len, version, y);
```

Comment:
i'm confused by this pattern, given that `get_async()` is async. shouldn't we wait for this block to complete before calling `update()` on it? the `completed` list here will only contain other blocks that previously completed - you should probably loop over those to make their corresponding `update()` calls

Reply:
@cbodley : This is on my TODO list - I have been thinking of shifting this to `flush` after the data is sent via `handle_data`. I need to figure out how to have the correct offset, length and version of the block accessible in `flush`. OR maybe simply do it after `flush`.

Reply:
`flush()` seems like a good place, yeah

> I need to figure out how to have the correct offset, length and version of the block

`completed` is a list of [rgw::AioResult](https://github.com/ceph/ceph/blob/5ea5521/src/rgw/rgw_aio.h#L34-L38)s. we put the offset in `AioResult::id` so we can sort the results. you can get the length from `AioResult::data.length()`. for version, i think that should be the same for all of the reads?

Reply:
Yes version shouldn't be a problem, but for ranged requests - the offset and length of the block cached might be different from the offset and length of data returned to the end user